### PR TITLE
Replace processors and Mockito.initMocks in azure-core-amqp

### PR DESCRIPTION
**Pull request by conniey**
* Replace instances of Direct, Replay, Unicast Processors with Sinks.
* Remove unneeded endpoint closed states. "complete" and "onError" are enough to signal terminal handlers.
* Replace `Mockito.initMocks` with openMocks

### CODE REVIEW

File: sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ActiveClientTokenManager.java [hunk 1]
```diff
@@ -85,7 +80,11 @@ public ActiveClientTokenManager(Mono<ClaimsBasedSecurityNode> cbsNode, String to
 
                     final Duration firstInterval = Duration.ofMillis(refreshIntervalMS);
                     lastRefreshInterval.set(firstInterval);
-                    authorizationResultsSink.next(AmqpResponseCode.ACCEPTED);
+                    authorizationResults.emitNext(AmqpResponseCode.ACCEPTED, (signalType, emitResult) -> {
+                        logger.info("signalType[{}] result[{}] Could not emit ACCEPTED.", signalType, emitResult);
```
Comment (alzimmermsft) [inline sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ActiveClientTokenManager.java:84]: How important are these emission, completion failures to the overall application? Is it ignorable or could it lead to downstream issues? What I'm wondering is if these failures should be logged at `info` or `verbose` depending on how often they occur and their importance.

File: sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorConnection.java [hunk 2]
```diff
@@ -48,9 +46,8 @@
     private final ConcurrentMap<String, SessionSubscription> sessionMap = new ConcurrentHashMap<>();
 
     private final AtomicBoolean isDisposed = new AtomicBoolean();
-    private final DirectProcessor<AmqpShutdownSignal> shutdownSignals = DirectProcessor.create();
-    private final FluxSink<AmqpShutdownSignal> shutdownSignalsSink = shutdownSignals.sink();
-    private final ReplayProcessor<AmqpEndpointState> endpointStates;
+    private final Sinks.One<AmqpShutdownSignal> shutdownSignalSink = Sinks.one();
+    private final Flux<AmqpEndpointState> endpointStates;
```
Comment (alzimmermsft) [inline sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorConnection.java:50]: Why was this changed to a `Flux` and not a `Sinks.many().replay().latest()`?

File: sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorConnection.java [hunk 3]
```diff
@@ -48,9 +46,8 @@
     private final ConcurrentMap<String, SessionSubscription> sessionMap = new ConcurrentHashMap<>();
 
     private final AtomicBoolean isDisposed = new AtomicBoolean();
-    private final DirectProcessor<AmqpShutdownSignal> shutdownSignals = DirectProcessor.create();
-    private final FluxSink<AmqpShutdownSignal> shutdownSignalsSink = shutdownSignals.sink();
-    private final ReplayProcessor<AmqpEndpointState> endpointStates;
+    private final Sinks.One<AmqpShutdownSignal> shutdownSignalSink = Sinks.one();
```
Comment (alzimmermsft) [inline sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorConnection.java:49]: Looking at the usage of this is there a `Sinks.ony().replay()`? So we don't need to `cache()` when converting to a `Flux`.

File: sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorReceiver.java [hunk 4]
```diff
@@ -38,9 +36,9 @@
     private final ReactorDispatcher dispatcher;
     private final Disposable subscriptions;
     private final AtomicBoolean isDisposed = new AtomicBoolean();
-    private final EmitterProcessor<Message> messagesProcessor;
+    private final Flux<Message> messagesProcessor;
```
Comment (alzimmermsft) [inline sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorReceiver.java:39]: Is there a Sink that could accomplish what we want here? From my understanding, and correct me if I'm wrong, but this is looking to cache messages until we have a subscriber and once a subscription happens emissions become hot and are only available at emission time and additional subscribers later only receive newly emitted messages.

File: sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorSender.java [hunk 5]
```diff
@@ -76,7 +75,7 @@
     private final PriorityQueue<WeightedDeliveryTag> pendingSendsQueue =
         new PriorityQueue<>(1000, new DeliveryTagComparator());
     private final ClientLogger logger = new ClientLogger(ReactorSender.class);
-    private final ReplayProcessor<AmqpEndpointState> endpointStates;
+    private final Flux<AmqpEndpointState> endpointStates;
```
Comment (alzimmermsft) [inline sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorSender.java:78]: I'm seeing this, `Flux<AmqpEndpointState>`, in many classes, is there a way we could extract this out into a single location where we instantiate it as a Sink?

File: sdk/core/azure-core-amqp/src/test/java/com/azure/core/amqp/implementation/ActiveClientTokenManagerTest.java [hunk 6]
```diff
@@ -49,7 +49,7 @@ static void afterAll() {
 
     @BeforeEach
     void setup() {
-        MockitoAnnotations.initMocks(this);
+        MockitoAnnotations.openMocks(this);
```
Comment (alzimmermsft) [inline sdk/core/azure-core-amqp/src/test/java/com/azure/core/amqp/implementation/ActiveClientTokenManagerTest.java:52]: `openMocks` returns a Closeable, we may want to capture that and close it in `AfterEach`

**Review by alzimmermsft — COMMENTED** [review]
Are we able to uncomment the configuration in `azure-core-amqp`'s POM file to fail build on deprecated API usage?

File: sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ActiveClientTokenManager.java (hunk 1 above)
Comment (conniey) [reply sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ActiveClientTokenManager.java:84]: Good call.. I have been trying to either remove or lower levels because some of them are rather verbose.

File: sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorConnection.java (hunk 2 above)
Comment (conniey) [reply sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorConnection.java:50]: Because this isn't programmatically emitted. On line 104, all of the results come from another flux. `handler.getEndpointStates()`.
https://github.com/Azure/azure-sdk-for-java/pull/19928/files/cfda5e9f016e9d9504f1826173e11b1e88a76d9e#diff-4fd2600dc96ee9531c22be07b35f51eab6a957073c9adff75d0c54440b617ec3R104

File: sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorConnection.java (hunk 3 above)
Comment (conniey) [reply sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorConnection.java:49]: Unfortunately there is no replay. Or I would have. :(

File: sdk/core/azure-core-amqp/src/test/java/com/azure/core/amqp/implementation/ActiveClientTokenManagerTest.java (hunk 6 above)
Comment (conniey) [reply sdk/core/azure-core-amqp/src/test/java/com/azure/core/amqp/implementation/ActiveClientTokenManagerTest.java:52]: Thanks for that catch, I'll do that!

File: sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorSender.java (hunk 5 above)
Comment (conniey) [reply sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorSender.java:78]: So handlers propagate up an endpoint state, proton-j `EndpointState`. Senders, receivers, sessions, and connections can all have different states... and based on each state, we want to either close child links, dispose deliveries, etc. So we subscribe to that handler endpoint, do some sort of processing to change it to an azure-core-amqp AmqpEndpointState and then surface that.

File: sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorReceiver.java (hunk 4 above)
Comment (conniey) [reply sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorReceiver.java:39]: We process messages by transforming them from the handler. There is no programmatic emission.

Comment (conniey) [reply sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorReceiver.java:39]: This was an incorrect use of processors before.

File: sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ActiveClientTokenManager.java (hunk 1 above)
Comment (srnagar) [reply sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ActiveClientTokenManager.java]: Yeah, making this verbose would be better as the user really can't diagnose much from this as it's mostly an implementation detail.

File: sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ActiveClientTokenManager.java [hunk 7]
```diff
@@ -99,8 +98,14 @@ public void close() {
...
         }
 
-        authorizationResultsSink.complete();
-        durationSourceSink.complete();
+        authorizationResults.emitComplete((signalType, emitResult) -> {
+            logger.info("signalType[{}] result[{}] Could not close authorizationResults.", signalType, emitResult);
+            return false;
+        });
+        durationSource.emitComplete((signalType, emitResult) -> {
+            logger.info("signalType[{}] result[{}] Could not close authorizationResults.", signalType, emitResult);
```
Comment (srnagar) [inline sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ActiveClientTokenManager.java]:
```suggestion
            logger.info("signalType[{}] result[{}] Could not close durationSource.", signalType, emitResult);
```

File: sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorReceiver.java [hunk 8]
```diff
@@ -66,14 +64,15 @@ protected ReactorReceiver(String entityPath, Receiver receiver, ReceiveLinkHandl
...
+            .publish()
+            .autoConnect();
         this.endpointStates = this.handler.getEndpointStates()
             .map(state -> {
                 logger.verbose("connectionId[{}], path[{}], linkName[{}]: State {}", handler.getConnectionId(),
                     entityPath, getLinkName(), state);
                 return AmqpEndpointStateUtil.getConnectionState(state);
             })
-            .subscribeWith(ReplayProcessor.cacheLastOrDefault(AmqpEndpointState.UNINITIALIZED));
+            .cache(1);
```
Comment (srnagar) [inline sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorReceiver.java:75]: cache(1) works a little differently than cacheLast which was previously used. `cache(n)` emits the last n items emitted by the source when a subscription is requested even if the source hasn't emitted a complete signal. This doesn't guarantee that the emitted signal is the last signal from the source.

Comment (conniey) [reply sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorReceiver.java:75]: I thought cacheLast was replay the last n items. I must have misunderstood it... phew. glad that was fixed.

File: sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorConnection.java (hunk 3 above)
Comment (alzimmermsft) [inline sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorConnection.java:49]: Based on https://projectreactor.io/docs/core/release/api/reactor/core/publisher/DirectProcessor.html the closest Sinks equivalent would be `Sinks.many().multicast().directBestEffort()`

File: sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/handler/ReceiveLinkHandler.java [hunk 9]
```diff
@@ -22,15 +21,12 @@
 public class ReceiveLinkHandler extends LinkHandler {
     private final String linkName;
     private final AtomicBoolean isFirstResponse = new AtomicBoolean(true);
-    private final DirectProcessor<Delivery> deliveries;
-    private final FluxSink<Delivery> deliverySink;
+    private final Sinks.Many<Delivery> deliveries = Sinks.many().multicast().onBackpressureBuffer();
```
Comment (alzimmermsft) [inline sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/handler/ReceiveLinkHandler.java:24]: Based on https://projectreactor.io/docs/core/release/api/reactor/core/publisher/DirectProcessor.html the closest Sinks equivalent would be `Sinks.many().multicast().directBestEffort()`

File: sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/handler/SendLinkHandler.java [hunk 10]
```diff
@@ -21,10 +19,8 @@
     private final String linkName;
     private final String entityPath;
     private final AtomicBoolean isFirstFlow = new AtomicBoolean(true);
-    private final UnicastProcessor<Integer> creditProcessor = UnicastProcessor.create();
-    private final DirectProcessor<Delivery> deliveryProcessor = DirectProcessor.create();
-    private final FluxSink<Integer> creditSink = creditProcessor.sink();
-    private final FluxSink<Delivery> deliverySink = deliveryProcessor.sink();
+    private final Sinks.Many<Integer> creditProcessor = Sinks.many().unicast().onBackpressureBuffer();
+    private final Sinks.Many<Delivery> deliveryProcessor = Sinks.many().multicast().onBackpressureBuffer();
```
Comment (alzimmermsft) [inline sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/handler/SendLinkHandler.java:23]: Based on https://projectreactor.io/docs/core/release/api/reactor/core/publisher/DirectProcessor.html the closest Sinks equivalent would be `Sinks.many().multicast().directBestEffort()`

File: sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorConnection.java (hunk 3 above)
Comment (conniey) [reply sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorConnection.java:49]: Yeah. There is ever going to be a single signal though. So since we had shipped the connection API, under the hood, it would be a single sink but exposed as that.

**Review by srnagar — APPROVED** [review]

File: sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/handler/ConnectionHandler.java [hunk 11]
```diff
@@ -156,8 +156,8 @@ protected void addTransportLayers(Event event, TransportInternal transport) {
 
     @Override
     public void onConnectionInit(Event event) {
-        logger.info("onConnectionInit hostname[{}], connectionId[{}], amqpHostname[{}]", getHostname(),
-            getConnectionId(), connectionOptions.getFullyQualifiedNamespace());
+        logger.info("onConnectionInit connectionId[{}] hostname[{}] amqpHostname[{}]",
+            getHostname(), getConnectionId(), connectionOptions.getFullyQualifiedNamespace());
```
Comment (YijunXieMS) [inline sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/handler/ConnectionHandler.java:160]: getHostname() should be after getConnectionId() to sync with the param order
